src/freenet/client/async/SplitFileInserterStorage.java

```diff
@@ -218,7 +223,16 @@
 
     private final int overallStatusLength;
     
+    /** Protects cooldown calculations and noBlocksToSend. You MUST NOT take this lock while 
+     * holding any other locks!!!
```

Comment:
Sorry, it's rather tangled... When we choose a block, if we can't find a block to send, we compute the time at which there will next be a block to send. This is the cooldown time. The request selection structures (RandomGrabArray etc) summarise this across groups of requests and thus avoid having to check the whole tree every time. Hence if something changes, e.g. a request completes with a DNF or we encode a segment, we need to clear the cache and then reschedule the request / tell the RGAs about it.

So we have concurrency issues: 
- Thread A fails to choose a block and sets cooldown to MAX_VALUE.
- Thread B completes a request and sets cooldown to 0
=> If they complete in the wrong order, the request will stall. So we need locking.

But we can't just use (this) because we need to check *all* the segments before we conclude that there are no blocks to send. So using (this) would probably cause deadlocks. Similarly the warnings about not taking cooldownLock while holding other locks - deadlocks could result, because when recomputing it we take cooldownLock and then the segment-level locks.

On top of all that, in getWakeupTime(), it'd be nice not to have to take (this) and the unlock it, and then take (cooldownLock) and unlock that. It *SHOULD* be safe to do this, which is what the first version of the patch did, because it's only run on the request selection thread. Overloading the cooldown time (just as it is overloaded in the getWakeupTime API) is an optimisation to avoid having to take two locks (one at a time). It's safe because we only go to "finished" once, and it's okay for there to be a minor delay between finishing and being removed from the RGA's. Also, it is supposed to be removed explicitly... possibly there's a concurrency issue there too...

The best solution would probably be to have a separate getter for each segment, and let the RGA's do the aggregation. That would require a larger refactoring and would use slightly more memory (and I understand people are already complaining about memory?).

Thoughts?